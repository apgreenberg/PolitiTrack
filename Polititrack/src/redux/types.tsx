import { UserInfo } from '../model/database/userInterfaces';
import { ThunkAction, ThunkDispatch } from 'redux-thunk';

import { store } from './store';

//Action types
export const LOGIN = 'LOGIN';
export const LOGOUT = 'LOGOUT';

export interface LoginAction {
  type: typeof LOGIN;
  payload: UserInfo;
}

export interface LogoutAction {
  type: typeof LOGOUT;
}

export type Actions = LoginAction | LogoutAction;

//State interface typing
export interface StateWrapper {
  loggedIn: boolean;
  currentUser: UserInfo;
}

//Dispatch types
export type ThunkActionType<ReturnType = void> = ThunkAction<
  ReturnType,
  StateWrapper,
  unknown,
  Actions
>;

export type ThunkDispatchType = ThunkDispatch<StateWrapper, undefined, Actions>;

export type AppDispatch = typeof store.dispatch;
