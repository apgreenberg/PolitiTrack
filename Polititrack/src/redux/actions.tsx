import { ThunkActionType } from './types';
import { ActionCreator, Dispatch } from 'redux';
import {
  attemptSignIn,
  attemptSignOut,
  attemptDeleteAccount,
  findUser,
  pullCurrentUserInfo,
} from '../model/database/handleLogin';
import { getNextId, addUser } from '../model/database/handleSignUp';
import { updateInternalDatabase } from '../model/database/userHelpers';
import { UserInfo } from '../model/database/userInterfaces';
import {
  initializeNotifications,
  removeNotifications,
} from '../model/background/notifications';

/**
 * Checks if the user exists and if so signs them in
 * @param email - the user's email
 * @param password - the user's password
 * @returns - An action to modify the logged in state
 */
export const signIn: ActionCreator<ThunkActionType> = (
  email: string,
  password: string
) => {
  return async (dispatch: Dispatch) => {
    const foundUserInfo: UserInfo | null = findUser(email, password);
    if (foundUserInfo !== null) {
      await attemptSignIn(foundUserInfo);
      await initializeNotifications(foundUserInfo);
      dispatch({ type: 'LOGIN', payload: foundUserInfo });
    } else {
      alert('User does not exist!');
    }
  };
};

/**
 * Signs out the current user
 * @returns - an action to modify the logged in state
 */
export const signOut: ActionCreator<ThunkActionType> = () => {
  return async (dispatch: Dispatch) => {
    await attemptSignOut();
    await removeNotifications();
    dispatch({ type: 'LOGOUT' });
  };
};

/**
 * Signs up a new user and logs them in
 * @param email - the new user's email
 * @param password - the new user's password
 * @returns - an action to modify the logged in state
 */
export const signUp: ActionCreator<ThunkActionType> = (
  email: string,
  password: string
) => {
  return async (dispatch: Dispatch) => {
    const id: number = getNextId();
    const newUser: UserInfo = {
      id: id,
      token: 'abc',
      email: email,
      password: password,
      name: 'default name',
      data: { jurisdictions: [], bills: [], billTypes: [], legislators: [] },
      settings: { updateJurisdictions: false, updateBills: false },
    };
    await addUser(newUser);
    await attemptSignIn(newUser);
    await initializeNotifications(newUser);
    dispatch({ type: 'LOGIN', payload: newUser });
  };
};

/**
 * Retrieves the past user from AsyncStorage if there is one
 * @returns - an action to modify the logged in state
 */
export const retrievePastUser: ActionCreator<ThunkActionType> = () => {
  return async (dispatch: Dispatch) => {
    await updateInternalDatabase();
    const userInfo: UserInfo | null = await pullCurrentUserInfo();
    if (userInfo !== null) {
      await initializeNotifications(userInfo);
      dispatch({ type: 'LOGIN', payload: userInfo });
    }
  };
};

/**
 * Logs out and deletes the user with the associated id
 * @param id - the user to delete
 * @returns - an action to modify the logged in state
 */
export const deleteAccount: ActionCreator<ThunkActionType> = (id: number) => {
  return async (dispatch: Dispatch) => {
    await attemptSignOut();
    await removeNotifications();
    await attemptDeleteAccount(id);
    dispatch({ type: 'LOGOUT' });
  };
};
