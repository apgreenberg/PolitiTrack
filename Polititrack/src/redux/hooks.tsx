import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux';
import { UserInfo } from '../model/database/userInterfaces';
import type { StateWrapper, AppDispatch } from './types';

export const useAppDispatch = () => useDispatch<AppDispatch>();
export const useAppSelector: TypedUseSelectorHook<StateWrapper> = useSelector;
export const selectLoginState = (state: StateWrapper): boolean =>
  state.loggedIn;
export const selectCurrentUser = (state: StateWrapper): UserInfo =>
  state.currentUser;
