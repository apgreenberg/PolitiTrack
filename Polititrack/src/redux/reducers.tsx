import { blankUserInfo } from '../model/database/userInterfaces';
import { Actions, LOGIN, LOGOUT, StateWrapper } from './types';

const initialState: StateWrapper = {
  loggedIn: false,
  currentUser: blankUserInfo,
};

const reducer = (
  state: StateWrapper = initialState,
  action: Actions
): StateWrapper => {
  switch (action.type) {
    case LOGIN:
      return {
        ...state,
        loggedIn: true,
        currentUser: action.payload,
      };
    case LOGOUT:
      return {
        ...state,
        loggedIn: false,
      };
    default:
      return state;
  }
};

export default reducer;
