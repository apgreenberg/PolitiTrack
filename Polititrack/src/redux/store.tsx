import { createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import reducer from './reducers';
import { ThunkDispatchType, StateWrapper } from './types';

export const store = createStore(
  reducer,
  applyMiddleware<ThunkDispatchType, StateWrapper>(thunk)
);
