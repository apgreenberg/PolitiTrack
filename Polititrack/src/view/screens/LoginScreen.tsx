import React from 'react';
import {
  Text,
  View,
  SafeAreaView,
  TextInput,
  Button,
  StyleSheet,
} from 'react-native';

import { useAppDispatch } from '../../redux/hooks';
import { signIn, signUp } from '../../redux/actions';
import { Dispatch, Action } from 'redux';
import { ThunkDispatchType } from '../../redux/types';

import text from '../styles/text';

/**
 * Creates the Login/Sign up pages
 * @returns - the Functional Components for the Login/Sign up page
 */
const LoginScreen: React.FC = () => {
  const dispatch: Dispatch<Action> & ThunkDispatchType = useAppDispatch();
  const [loginData, setLoginData] = React.useState({
    email: '',
    password: '',
  });

  return (
    <SafeAreaView style={styles.container}>
      <Text style={text.mediumText}>Email:</Text>
      <TextInput
        style={styles.inputBoxes}
        onChangeText={(text) => setLoginData({ ...loginData, email: text })}
      />
      <Text style={text.mediumText}>Password:</Text>
      <TextInput
        style={styles.inputBoxes}
        secureTextEntry={true}
        onChangeText={(text) => setLoginData({ ...loginData, password: text })}
      />
      <View style={{ flexDirection: 'row', alignItems: 'center' }}>
        <Button
          testID={'loginButton'}
          title="Login"
          onPress={() => {
            dispatch(signIn(loginData.email, loginData.password));
          }}
        />
        <Text style={text.mediumText}> or </Text>
        <Button
          testID={'signUpButton'}
          title="Sign Up"
          onPress={() => {
            dispatch(signUp(loginData.email, loginData.password));
          }}
        />
      </View>
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
  },
  inputBoxes: {
    borderWidth: 2,
    width: '60%',
    height: 40,
    fontSize: 20,
  },
});
export default LoginScreen;
