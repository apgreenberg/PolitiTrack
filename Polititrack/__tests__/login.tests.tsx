import React from 'react';
import TestRenderer from 'react-test-renderer';
import { render, fireEvent } from '@testing-library/react-native';
import LoginScreen from '../src/view/screens/LoginScreen';
import * as actions from '../src/redux/actions';
const signInMock = jest.spyOn(actions, 'signIn');
const signUpMock = jest.spyOn(actions, 'signUp');

/**
 * Tests the login screen, pressing the sign in and sign up buttons
 */
describe('Login tests', () => {
  afterEach(() => {
    jest.resetModules();
  });
  test('Mock user has not been logged in', () => {
    const { toJSON } = render(<LoginScreen />);
    expect(toJSON()).toMatchSnapshot();
  });

  test('Test pressing the login button', () => {
    const { getByTestId } = render(<LoginScreen />);
    fireEvent(getByTestId('loginButton'), 'onPress');
    expect(signInMock).toHaveBeenCalled();
  });
  test('Test pressing the sign up button', () => {
    const { getByTestId } = render(<LoginScreen />);
    fireEvent(getByTestId('signUpButton'), 'onPress');
    expect(signUpMock).toHaveBeenCalled();
  });
});
