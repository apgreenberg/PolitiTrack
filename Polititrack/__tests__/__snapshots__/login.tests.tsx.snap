// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login tests Mock user has not been logged in 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Text
    style={
      Object {
        "fontSize": 16,
        "textAlign": "center",
      }
    }
  >
    Email:
  </Text>
  <TextInput
    allowFontScaling={true}
    onChangeText={[Function]}
    rejectResponderTermination={true}
    style={
      Object {
        "borderWidth": 2,
        "fontSize": 20,
        "height": 40,
        "width": "60%",
      }
    }
    underlineColorAndroid="transparent"
  />
  <Text
    style={
      Object {
        "fontSize": 16,
        "textAlign": "center",
      }
    }
  >
    Password:
  </Text>
  <TextInput
    allowFontScaling={true}
    onChangeText={[Function]}
    rejectResponderTermination={true}
    secureTextEntry={true}
    style={
      Object {
        "borderWidth": 2,
        "fontSize": 20,
        "height": 40,
        "width": "60%",
      }
    }
    underlineColorAndroid="transparent"
  />
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
      }
    }
  >
    <View
      accessibilityRole="button"
      accessibilityState={Object {}}
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID="loginButton"
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "margin": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          Login
        </Text>
      </View>
    </View>
    <Text
      style={
        Object {
          "fontSize": 16,
          "textAlign": "center",
        }
      }
    >
       or 
    </Text>
    <View
      accessibilityRole="button"
      accessibilityState={Object {}}
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID="signUpButton"
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "margin": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          Sign Up
        </Text>
      </View>
    </View>
  </View>
</RCTSafeAreaView>
`;
